### ISSUE TITLE
aria- and form input errors

### ISSUE THREAD

**Issue by joshbruce** [18F/web-design-standards#1145]
From the html5 spec:

http://www.w3.org/TR/wai-aria/states_and_properties#aria-describedby

>Identifies the element (or elements) that describes the object. See related aria-labelledby.
>
>The aria-labelledby attribute is similar to aria-describedby in that both reference other elements to calculate a text alternative, but a label should be concise, where a description is intended to provide more verbose information.
>
>The element or elements referenced by the aria-describedby comprise the entire description. Include ID references to multiple elements if necessary, or enclose a set of elements (e.g., paragraphs) with the element referenced by the ID.

Thinking it might be worthwhile to shift the error inputs to `aria-labelledby` - leaving the option for an `aria-describedby` - say instructional documentation, which, I believe, should live on the same page according to the spec.

**Comment by msecret** [comment]
@joshbruce thanks for the issue, I'm checking this with our accessibility experts to hear what they have to say and will get back to you soon.

**Comment by KimberlyMunoz** [comment]
I'm not sure if either is discoverable unless you do focus management to make sure to move the focus to the errors after submit. 

Whether you use `aria-labelledby` or `aria-describedby` (which I'd want to test to make sure `aria-labelledby` doesn't conflict with `label` in most standard screenreaders) appropriate focus management might be something worth adding to the guidelines.

**Comment by nickbristow** [comment]
`aria-labelledby` overwrites the default label. You can also do `aria-labelledby='for-value-of-lable{SPACE}id-of-instructions'` so that it reads both/multiple things.

`aria-describedby` should read additional info, but may not be read by default depending on your screen reader.

**Comment by joshbruce** [comment]
To verify I am understanding.

In the this example, the label is read and considered the `label`:

```
<label for="input-id">Hello</label>
<input id="input-id">
```

In this example, the `aria-labelledby` causes the `label` to be overridden by the `span`:

```
<label for="input-id">Hello</label>
<span id="input-id-error-message">Helpful text</span>
<input id="input-id" aria-labelledby="input-id-error-message">
```

Both the following *could* result in a similar output from the screen reader, but the `aria-describedby` may not be read by default:


```
<label for="input-id">Hello</label>
<span id="input-id-error-message">Helpful text</span>
<input id="input-id" aria-labelledby="input-id-error-message input-id-long-description">
<p id="input-id-long-description">Some longer instructions.</p>
```

- or -

```
<label for="input-id">Hello</label>
<span id="input-id-error-message">Helpful text</span>
<input id="input-id" aria-labelledby="input-id-error-message" aria-describedby="input-id-long-description">
<p id="input-id-long-description">Some longer instructions.</p>
```

In all cases, except the first, the `label` is overridden as the descriptor for the element?

**Comment by nickbristow** [comment]
@joshbruce So I put some examples on our little guide page. If you have a screen reader you can test these examples. I also recorded the output of voiceover there as well. 

The gist is, `aria-label`, `aria-labeledby` override the default label and `aria-describedby` doesn't work with voice over on OSX but should work with jaws. 

https://pages.18f.gov/accessibility/forms/#car_1

**Comment by joshbruce** [comment]
@nickbristow - Thank! I've been toying with VoiceOver a bit over the last couple of days. I haven't had a chance to test against this issue, and definitely appreciate the assist. That's too bad that VoiceOver doesn't recognize the `aria-describedby`.

Thinking about testing an internal link to accomplish the same - for both those using assistive technology and those not. 

Thanks again for following up.

**Comment by nickbristow** [comment]
You're welcome!

On Friday, May 20, 2016, Josh Bruce <notifications@github.com> wrote:

> @nickbristow <https://github.com/nickbristow> - Thank! I've been toying
> with VoiceOver a bit over the last couple of days. I haven't had a chance
> to test against this issue, and definitely appreciate the assist. That's
> too bad that VoiceOver doesn't recognize the aria-describedby.
>
> Thinking about testing an internal link to accomplish the same - for both
> those using assistive technology and those not.
>
> Thanks again for following up.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/18F/web-design-standards/issues/1145#issuecomment-220723089>
>


-- 
Nick Bristow
18F Innovation Specialist
Nick.Bristow@gsa.gov
